[HUDI-5753] Add docs for record payload

codope opened the pull request:
### Change Logs

Documentation about record payload under `Concepts` section.

### Impact

Only docs.

### Risk level (write none, low medium or high below)

low

### Documentation Update

Adds a new page under `Concepts` section of Hudi website.

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

nsivabalan commented on website/docs/record_payload.md:
```diff
@@ -0,0 +1,97 @@
+#### OverwriteWithLatestAvroPayload
+
+This is the default record payload implementation. It picks the record with the greatest value (determined by calling
+.compareTo() on the value of precombine key) to break ties and simply picks the latest record while merging. This gives
```
actually there is little more to this. lets land this doc for 0.13.0. but as an immediate follow up, address these comments. 
we have precombine and combineAndGetUpdate method used in diff occasions. 
so calling out just preCombine may not be right. bcoz, when merging w/ whats in storage, we ignore the preCombine value specifically in this payoad(OverwriteWithLatestAvroPayload)

codope replied:
That's true. But, I wanted to keep things simple for the user as it is a concepts doc. Towards the end, I have pointed to the FAQ which has more details.

nsivabalan commented on website/docs/record_payload.md:
```diff
@@ -0,0 +1,97 @@
+
+#### ExpressionPayload
+
+This payload is very useful when you want to merge or delete records based on some conditional expression, especially
```
should we remove this from this list. I thought its meant to be used only internally. can anyone directly set expression payload for their table?

codope replied:
Didn't know that this is meant to be used internally. Is there a guard like that on payload class config? cc @alexeykudinkin

alexeykudinkin replied:
Correct this is only supposed to be used internally (by MERGE INTO)

nsivabalan commented on website/docs/record_payload.md:
```diff
@@ -0,0 +1,97 @@
+There are quite a few other implementations provided by Hudi. For example, `MySqlDebeziumAvroPayload` and
+`PostgresDebeziumAvroPayload` provides support for seamlessly applying changes captured via Debezium for MySQL and
+PostgresDB.
+`AWSDmsAvroPayload` provides support for applying changes captured via Amazon Database Migration Service onto S3.
```
OverwriteNonDefaultsWithLatestAvroPayload, DefaultHoodieRecordPayload

nsivabalan commented on website/docs/record_payload.md:
```diff
@@ -0,0 +1,97 @@
+
+#### EventTimeAvroPayload
+
+Some use cases require merging records by event time and thus event time plays the role of an ordering field. This
```
curious to know, how is this diff from using DefaultHoodieRecordPayload where we use the event time as the payload ordering field.

codope replied:
Not much. `DefaultHoodieRecordPayload` maintains some additional metadata to track latency and freshness. I wanted to write about `DefaultHoodieRecordPayload` but the naming belies the actual default. So, I avoided confusion. Perhaps, we should make this the actual default. Is it covered in config simplification story? cc @bhasudha 
Let me just add some notes about `DefaultHoodieRecordPayload`.

nsivabalan commented on website/docs/record_payload.md:
````diff
@@ -0,0 +1,97 @@
+
+```
+    id      ts      name    price
+    1       2       name_1  price_1
````
how is ts's value is 2? its not intuitive to me. I thought, for null values in new incoming, we will choose older value.

codope replied:
`ts` is the ordering field so the record with higher value is picked. Null value for `name` column in incoming record indeed gets replaced by value in the existing record.

nsivabalan replied:
this is bit confusing actually. 
so based on ordering field, we choose the one in storage. so, shouldn't we ignore the new record totally?

nsivabalan replied:
oh I see. if the two records came in proper order, the final snapshot would have been 
1 2 name_1 price_1

I get it now.

nfarah86 commented:
Following up- I need to use this doc for the write operations, so I can reference the payloads under the UPSERT operation.

nfarah86 commented:
@codope is it possible you can provide an example to extend the payload for a customized option. Also, are there configs the user should consider that's provided out-of-the-box. If possible, can you specify all of them inline with the right class?

codope commented:
> @codope is it possible you can provide an example to extend the payload for a customized option. Also, are there configs the user should consider that's provided out-of-the-box? If possible, can you specify all of them inline with the right class?

@nfarah86 I have added a link to FAQ where there are more details on how to implement a custom payload. I have also removed the record merger API. Need to follow up with a separate doc or update this doc in a separate PR.
